riptide-spring-boot-autoconfigure/src/main/java/org/zalando/riptide/autoconfigure/OpenTracingFlowIdAutoConfiguration.java

```diff
@@ -0,0 +1,55 @@
+package org.zalando.riptide.autoconfigure;
+
+import io.opentracing.Tracer;
+import org.apache.hc.client5.http.classic.HttpClient;
+import org.apiguardian.api.API;
+import org.springframework.boot.autoconfigure.AutoConfiguration;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.zalando.opentracing.flowid.Flow;
+
+import static org.apiguardian.api.API.Status.INTERNAL;
+import static org.apiguardian.api.API.Status.STABLE;
+
+/**
+ * This class is adapted from org.zalando.opentracing.flowid.autoconfigure.OpenTracingFlowIdAutoConfiguration.
+ * The original class is based on org.zalando.opentracing.flowid.httpclient.FlowHttpRequestInterceptor,
+ * which uses Apache HttpClient 4.x.
+ * We use org.zalando.riptide.autoconfigure.FlowHttpRequestInterceptor, that is adapted to Apache HttpClient 5.x.
+ */
+@API(status = STABLE)
+@AutoConfiguration
+@ConditionalOnBean(Tracer.class)
+@AutoConfigureAfter(name = {
+        "org.springframework.boot.autoconfigure.web.servlet.WebMvcAutoConfiguration",
+        "io.opentracing.contrib.spring.web.starter.ServerTracingAutoConfiguration"
+})
+public class OpenTracingFlowIdAutoConfiguration {
+
+    @API(status = INTERNAL)
+    @Bean
+    public Flow flow(final Tracer tracer) {
+        return Flow.create(tracer);
+    }
+
+    @API(status = INTERNAL)
+    @Configuration
+    @ConditionalOnClass(HttpClient.class)
+    @ConditionalOnMissingBean(FlowHttpRequestInterceptor.class)
+    @ConditionalOnProperty(name = "opentracing.flowid.httpclient.enabled", havingValue = "true", matchIfMissing = true)
```

Comment:
matchIfMissing option should be false here.

But generally I'm concerned that http client tries now to assume that it's working in http servlet environment and establishing classes on incoming layer of the app.

Reply:
> matchIfMissing option should be false here.

The tests won't pass if it's set to false - and I guess it will break existing Riptide users because they may not have set `opentracing.flowid.httpclient.enabled` by default.

We can instead enable it with either `riptide.defaults.tracing.enabled = true` or `riptide.clients.*.tracing.enabled = true` being true (the latter having a wildcard makes it a bit hard 🤔).

> But generally I'm concerned that http client tries now to assume that it's working in http servlet environment and establishing classes on incoming layer of the app.

To my understanding, it doesn't make this assumption. However, if it works in such environment, Spring gets the hint how to order auto-configurations.

The current version of Riptide works in the same way - it just imports this class from `opentracing-flowid-starter`:
https://github.com/zalando/opentracing-toolbox/blob/main/opentracing-flowid/opentracing-flowid-autoconfigure/src/main/java/org/zalando/opentracing/flowid/autoconfigure/OpenTracingFlowIdAutoConfiguration.java

Since that project is in maintenance mode, I thought the only way to keep OpenTracing is add this little configuration.

Reply:
I think, it was always a bit weird that the FlowId topic was linked to Open Tracing. But I think, this was the issue when the original tracer project was started.

While it has something to do with tracing in general, in the end it is `just` an interceptor fiddling with the `X-Flow-Id` header. So given that we want to release a major version, we should think about whether it makes sense to decouple it and just use the `tracing.propagate-flow-id` property for it?

Reply:
I agree that it can be an improvement to decouple it, but it falls out the scope of this PR, which is "just" migration to SB 3.😊 

We can have another task in the milestone of Release 4.0 to apply the suggested decoupling. The rationale behind not doing so is to give this PR a chance to being released as soon as possible.

With Logbook, we did multiple release candidates, used them in our services, found the issues, and fixed them until we could make the new major release. I can imaginge we'll do the same, and in the meantime we have the chance to work on a decopling.